fsm/fsm.go

```diff
@@ -57,64 +179,88 @@ type FSM struct {
-
-	previousState := m.State
-
-	tupleFromAny := CmdStateTuple{cmd, -1}
-	tupleNormal := CmdStateTuple{cmd, m.State}
-	tupleCmdAny := CmdStateTuple{"any", m.State}
-
-	if fsmDomain.TransitionTable[tupleFromAny] == nil {
-		if fsmDomain.TransitionTable[tupleCmdAny] == nil {
-			transition = fsmDomain.TransitionTable[tupleNormal] // There is no transition "From Any" with cmd, nor "Cmd Any"
-			tuple = tupleNormal
-		} else {
-			transition = fsmDomain.TransitionTable[tupleCmdAny] // There is a transition "Cmd Any"
-			tuple = tupleCmdAny
+// TODO: Document what the ExecuteCmd does
+func (m *FSM) ExecuteCmd(cmd, question string, fsmDomain *Domain) (answers []query.Answer, extension string) {
+	cmdStateTuple, transitionFunc := m.SelectStateTransition(cmd, fsmDomain)
+
+	// Save information from the user's input into the slot
+	m.SetSlot(fsmDomain.SlotTable[cmdStateTuple], question)
+
+	// Transition FSM state and get answers or extension to execute
+	return m.TransitionState(cmd, transitionFunc, fsmDomain.DefaultMessages)
+}
+
+// SelectStateTransition based on the command provided
+func (m *FSM) SelectStateTransition(cmd string, fsmDomain *Domain) (CmdStateTuple, TransitionFunc) {
+	// fromAnyState means we can transition from any state
+	fromAnyState := CmdStateTuple{cmd, -1}
+
+	// normalState means we can transition from one existing state to another
+	normalState := CmdStateTuple{cmd, m.State}
+
+	// TODO: Whats the difference between fromAnyState and cmdAny?
+	cmdAnyState := CmdStateTuple{"any", m.State}
+
+	log.WithField("type", "fsm").Info(spew.Sprint(m))
+	log.WithField("type", "domain").Info(spew.Sprint(fsmDomain))
+
+	if fsmDomain.TransitionTable[fromAnyState] == nil {
+		if fsmDomain.TransitionTable[cmdAnyState] == nil {
+			// There is no transition "From Any" with cmd, nor "Cmd Any"
+			return normalState, fsmDomain.TransitionTable[normalState]
 		}
-	} else {
-		transition = fsmDomain.TransitionTable[tupleFromAny] // There is a transition "From Any" with cmd
-		tuple = tupleFromAny
+
+		// There is a transition "Cmd Any"
+		return cmdAnyState, fsmDomain.TransitionTable[cmdAnyState]
 	}
 
-	slot := fsmDomain.SlotTable[tuple]
+	// There is a transition "From Any" with cmd
+	return fromAnyState, fsmDomain.TransitionTable[fromAnyState]
+}
 
-	// Get slots
-	if slot.Name != "" {
-		switch slot.Mode {
-		case "whole_text":
-			m.Slots[slot.Name] = txt
+// SetSlot saves information from the user's input/question
+func (m *FSM) SetSlot(slot Slot, question string) {
+	slotName := strings.TrimSpace(slot.Name)
+	slotRegex := strings.TrimSpace(slot.Regex)
+
+	if slotName != "" {
+		switch strings.TrimSpace(slot.Mode) {
 		case "regex":
-			if r, err := regexp.Compile(slot.Regex); err == nil {
-				match := r.FindAllString(txt, 1)
+			if r, err := regexp.Compile(slotRegex); err == nil {
+				match := r.FindAllString(question, 1)
 				if len(match) > 0 {
-					m.Slots[slot.Name] = match[0]
+					m.Slots[slotName] = match[0]
 				}
 			}
+		default:
```

Comment:
Added a default case here. Which makes `whole_text` the default matching type.

Reply:
Neat!